### PR TITLE
chore(ci): build and test packaged app on the newer macos version (2nd attempt) COMPASS-8090

### PR DESCRIPTION
**Pull request by kraenhansen**
## Description

Second attempt at https://github.com/mongodb-js/compass/pull/6220 (which had to be reverted in #6230). It was reverted because running the tests were not `patchable` and failed because the host was not logged into a graphical shell.

Here's a patch which sets `patchable: true` to convince yourself this works: https://spruce.mongodb.com/version/66fedc4f7d7d0700078cc466/tasks

A friendly takeover of https://github.com/mongodb-js/compass/pull/5990 incorporating @mcasimir's feedback: https://github.com/mongodb-js/compass/pull/5990#discussion_r1662392749

Updates the build variants used for packaging the app on MacOS to 14 and adds two new test variants for the MacOS 14. We might want to simply remove `test-packaged-app-macos-11-arm` and `test-packaged-app-macos-11-x64` in the future.

Note: To avoid this regression in the future, I experimented with added a task taking a screenshot as an early sanity check. While this worked on the Arm64 Mac 14s, this failed on the `macos-1100` hosts (likely due to lack of permissions) although tests run perfectly fine.

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: .evergreen/buildvariants-and-tasks.in.yml [hunk 1]
```diff
@@ -443,6 +455,7 @@ tasks:
       - assigned_to_jira_team_compass_compass
       - foliage_check_task_only
     commands:
+      - func: screencapture
```
Comment (kraenhansen) [inline .evergreen/buildvariants-and-tasks.in.yml:458]: Adding a call to a new `screencapture` function here to exit early on MacOS if the host isn't logged into a graphical shell.

File: .evergreen/functions.yml [hunk 2]
```diff
@@ -247,6 +247,33 @@ functions:
...
+        shell: bash
+        script: |
+          set -e
+
+          die() {
+            echo $1 >&2
+            exit 1
+          }
+
+          if [[ $(uname) == Darwin ]]; then
```
Comment (kraenhansen) [inline .evergreen/functions.yml:264]: Performing the screencapture on MacOS only for now.

File: packages/compass-e2e-tests/helpers/compass.ts [hunk 3]
```diff
@@ -644,6 +645,15 @@ async function startCompassElectron(
...
   }
 
+  // Logging output from Electron, even before the app loads any JavaScript
+  const electronLogFile = path.join(LOG_PATH, `electron-${runCounter}.log`);
+  chromeArgs.push(
+    // See https://www.electronjs.org/docs/latest/api/command-line-switches#--enable-loggingfile
+    '--enable-logging=file',
+    // See https://www.electronjs.org/docs/latest/api/command-line-switches#--log-filepath
+    `--log-file=${electronLogFile}`
+  );
```
Comment (kraenhansen) [inline packages/compass-e2e-tests/helpers/compass.ts:655]: Capturing electron logs into a file even before the JS logging is initialized.

File: .evergreen/functions.yml [hunk 4]
```diff
@@ -247,6 +247,33 @@ functions:
...
+            screencapture ./screencapture.png || die "Screen capturing failed and it's likely that the host is not logged into a graphical shell. Post this in #ask-devprod-evergreen or create a 'Runtime Environments' DEVPROD ticket on Jira."
```
Comment (kraenhansen) [inline .evergreen/functions.yml:265]: Exiting with an actional error message if this fails.

**Review by lerouxb — APPROVED** [review]

**Pull request merged**